networking_aci/extensions/acioperations.py

```diff
@@ -0,0 +1,219 @@
+        return dict(hostgroup_name=hostgroup_name, hostgroup_mode=self._get_mode(ctx, hostgroup_name))
+
+    def update(self, request, **kwargs):
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        new_mode = kwargs.get("body", {}).get("mode")
+        if new_mode not in (aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL):
+            raise web_exc.HTTPBadRequest("Specify a mode that is either {} or {} (given mode was '{}')"
+                                         .format(aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL, new_mode))
+
+        ctx = context.get_admin_context()
+        curr_mode = self._get_mode(ctx, hostgroup_name)
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
+        hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+        if not hg_config:
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
+                    raise web_exc.HTTPConflict("Could not switch, there is still a portbinding for host {} "
+                                               "present in segment {}"
+                                               .format(host, segment_id))
+
+        # cleanup old baremetal objects when switching away from baremetal mode
+        baremetal_bindings_cleaned = False
+        if hg_config['hostgroup_mode'] == aci_const.MODE_BAREMETAL:
+            LOG.info("Cleaning baremetal objects for %s from device", hostgroup_name)
+            try:
+                self.rpc_notifier.clean_baremetal_objects(hg_config)
+                baremetal_bindings_cleaned = True
+            except Exception:
+                LOG.exception("Could not clean baremetal bindings from ACI, cleaning will have to be done manually")
+
+        if self.db.set_hostgroup_mode(ctx, hostgroup_name, new_mode):
+            LOG.info("Hostgroup %s set to mode %s", hostgroup_name, new_mode)
+
+            hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+            aci_objects_update_succeeded = False
+            try:
+                # switching policy group of port selectors, etc.
+                self.rpc_notifier.sync_direct_mode_config(hg_config)
+                aci_objects_update_succeeded = True
+            except Exception:
+                LOG.exception("Could not update bindings on ACI, rpc call failed")
+
+            return {
+                "success": True,
```

Comment:
We always return non-JSON error messages when something broke along the line, so I would assume a status-code of 200 would be enough for success and this additional key is basically useless, because it can only be there, if we succeeded.

Reply:
I added this for the case that someone using the API would like to check for success in a `if "success" in req.json()` manner. Do you think this is a bad idea?

Reply:
Not necessarily. I'd prefer if they'd check the HTTP status-code though.
Doesn't that type of pattern break for every error we return or does Neutron convert those errors into JSON, too?

Reply:
Neutron converts (most) errors to JSON as well